src/callouts.rs

```diff
@@ -0,0 +1,354 @@
+use anyhow::{anyhow, Context, Result};
+use libsystemd::logging::{self, Priority};
+use log::{debug, warn};
+use std::ffi::OsStr;
+use std::fmt::{self, Display, Formatter};
+use std::io::Write;
+use std::path::{Path, PathBuf};
+use std::process::{Command, Output, Stdio};
+
+use crate::mdev::*;
+
+#[derive(Clone, Copy, PartialEq)]
+pub enum Event {
+    Pre,
+    Post,
+    Notify,
+    Get,
+}
+
+impl Display for Event {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            Event::Pre => {
+                write!(f, "pre")
+            }
+            Event::Post => {
+                write!(f, "post")
+            }
+            Event::Notify => {
+                write!(f, "notify")
+            }
+            Event::Get => {
+                write!(f, "get")
+            }
+        }
+    }
+}
+
+#[derive(Clone, Copy)]
+pub enum Action {
+    Start,
+    Stop,
+    Define,
+    Undefine,
+    Modify,
+    Attributes,
+}
+
+impl Display for Action {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            Action::Start => write!(f, "start"),
+            Action::Stop => write!(f, "stop"),
+            Action::Define => write!(f, "define"),
+            Action::Undefine => write!(f, "undefine"),
+            Action::Modify => write!(f, "modify"),
+            Action::Attributes => write!(f, "attributes"),
+        }
+    }
+}
+
+#[derive(Clone, Copy)]
+pub enum State {
+    None,
+    Success,
+    Failure,
+}
+
+impl Display for State {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            State::None => write!(f, "none"),
+            State::Success => write!(f, "success"),
+            State::Failure => write!(f, "failure"),
+        }
+    }
+}
+
+pub struct Callout {
+    state: State,
+    script: Option<PathBuf>,
+    use_syslog: bool,
+}
+
+impl Callout {
+    fn new() -> Callout {
+        Callout {
+            state: State::None,
+            script: None,
+            use_syslog: false,
+        }
+    }
+
+    pub fn invoke<F>(dev: &mut MDev, use_syslog: bool, action: Action, func: F) -> Result<()>
+    where
+        F: Fn(&mut MDev) -> Result<()>,
+    {
+        let mut c = Callout::new();
+
+        if use_syslog {
+            c.set_use_syslog(use_syslog);
+        }
+
+        let res = match c.callout(dev, Event::Pre, action) {
+            Ok(_) => {
+                let tmp_res = func(dev);
+                c.state = match tmp_res {
+                    Ok(_) => State::Success,
+                    Err(_) => State::Failure,
+                };
+
+                let post_res = c.callout(dev, Event::Post, action);
+                if post_res.is_err() {
+                    debug!("Error occurred when executing post callout script");
+                }
+
+                tmp_res
+            }
+            Err(e) => Err(e),
+        };
```

Comment:
Any time you have a branch that just maps from the input to an identical output (`Err(e) => Err(e)`), there's probably a better way to do things. In this case, I'd suggest something like this:

```suggestion
        let res = c.callout(dev, Event::Pre, action).and_then(|_| {
            let tmp_res = func(dev);
            c.state = match tmp_res {
                Ok(_) => State::Success,
                Err(_) => State::Failure,
            };

            let post_res = c.callout(dev, Event::Post, action);
            if post_res.is_err() {
                debug!("Error occurred when executing post callout script");
            }

            tmp_res
        });

```